source/extensions/filters/http/cache/http_cache_utils.cc

```diff
@@ -0,0 +1,175 @@
+
+#include <array>
+#include <string>
+
+#include "absl/algorithm/container.h"
+#include "absl/strings/ascii.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/strip.h"
+
+using absl::ascii_isalnum;
+using absl::ascii_isdigit;
+using absl::Duration;
+using absl::InfiniteDuration;
+using absl::InfinitePast;
+using absl::ParseTime;
+using absl::Seconds;
+using absl::SimpleAtoi;
+using absl::string_view;
+using absl::Time;
+using absl::ZeroDuration;
+using Envoy::Http::HeaderEntry;
+using std::array;
+using std::string;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace Internal {
+
+namespace {
+
+bool tchar(char c) {
+  switch (c) {
+  case '!':
+  case '#':
+  case '$':
+  case '%':
+  case '&':
+  case '*':
+  case '+':
+  case '-':
+  case '.':
+  case '^':
+  case '_':
+  case '`':
+  case '|':
+  case '~':
+    return true;
+  }
+  return ascii_isalnum(c);
+}
+
+bool eatToken(string_view& s) {
+  const string_view::iterator token_end = c_find_if_not(s, &tchar);
+  if (token_end == s.begin()) {
+    return false;
+  }
+  s.remove_prefix(token_end - s.begin());
+  return true;
+}
+
+void eatDirectiveArgument(string_view& s) {
+  if (s.empty()) {
+    return;
+  }
+  if (s.front() == '"') {
+    // TODO(toddmgreer) handle \-escaped quotes
+    const size_t closing_quote = s.find('"', 1);
+    s.remove_prefix(closing_quote);
+  } else {
+    eatToken(s);
+  }
+}
+} // namespace
+
+// If s is nonnull and begins with decimal digits, return Eat leading digits in
+// *s, if any
+Duration eatLeadingDuration(string_view* s) {
+  const string_view::iterator digits_end = c_find_if_not(*s, &ascii_isdigit);
+  const size_t digits_length = digits_end - s->begin();
+  if (digits_length == 0) {
+    return ZeroDuration();
+  }
+  const string_view digits(s->begin(), digits_length);
+  s->remove_prefix(digits_length);
+  uint64_t num;
+  return SimpleAtoi(digits, &num) ? Seconds(num) : InfiniteDuration();
+}
+
+Duration effectiveMaxAge(string_view cache_control) {
+  // The grammar for This Cache-Control header value should be:
+  // Cache-Control   = 1#cache-directive
+  // cache-directive = token [ "=" ( token / quoted-string ) ]
+  // token           = 1*tchar
+  // tchar           = "!" / "#" / "$" / "%" / "&" / "'" / "*" / "+"
+  //                 / "-" / "." / "^" / "_" / "`" / "|" / "~" / DIGIT / ALPHA
+  // quoted-string   = DQUOTE *( qdtext / quoted-pair ) DQUOTE
+  // qdtext          = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
+  // obs-text        = %x80-FF
+  // quoted-pair     = "\" ( HTAB / SP / VCHAR / obs-text )
+  // VCHAR           =  %x21-7E  ; visible (printing) characters
+  Duration max_age = -InfiniteDuration();
+  bool found_s_maxage = false;
+  while (!cache_control.empty()) {
+    // Each time through the loop, we eat one cache-directive. Each branch
+    // either returns or completely eats a cache-directive.
+    if (ConsumePrefix(&cache_control, "no-cache")) {
```

Comment:
I've parsed cache-control in the past by first splitting on "," and trimming spaces. Up to you though.  This just seems a little more worrisome that you might walk off the end of the string.

My higher level comment is that you've done all this parsing work and only extracted one of the many things you need to know about cache-control. I feel like it's worth having a class for this which does some parsing and populates whatever you need to learn about the caching headers.

Reply:
The problems with splitting on ',' is that it makes it harder to correctly handle (by ignoring) absurd-but-legal cache-control-directives like 'public, foo="I like commas, do you", max-age=99', and that it necessitates a second pass over each directive. It isn't clear that either issue really matters, so I could be convinced to switch approaches for the sake of easier to understand code.

I do expect to turn this into a CacheControl class at some point, but wanted to get the most basic form out first.

Reply:
Good point about dealing with quoted cache-control elements; I didn't think there were any though; seems evil.  In any case what you have is readable and able to deal with that a little better, so it's fine.

But can you add a TODO about the CacheControl class? It might be something to put in source/common/http and shared with the gzip filter which examines cache-control too.

Reply:
The "private" directive's argument is supposed to be quoted if present, and extension directives are permitted to have quoted arguments. That's reasonable, but if someone puts a ',' in one of those quoted strings, I'd bet money that it's either a bug or an attack--I'll consider whether it would be use ','-splitting and then reject split quotes.

TODO added.